### PR TITLE
Add binary_and_multiclass

### PR DESCRIPTION
**Pull request by Eleven1Liu**
**Description**
- **Add linear solver**: `binary_and_multiclass`
- **linear.prepare_options**: check `-s`  > 7

### CODE REVIEW

File: libmultilabel/linear/linear.py [hunk 1]
```diff
@@ -66,23 +68,29 @@ def prepare_options(x: sparse.csr_matrix, options: str) -> 'tuple[sparse.csr_mat
     if any(o in options for o in ['-R', '-C', '-v']):
         raise ValueError('-R, -C and -v are not supported')
 
+    options_split = options.split()
+    if '-s' in options_split:
+        i = options_split.index('-s')
+        solver_type = int(options_split[i+1])
+        if solver_type >= 11:
+            raise ValueError("Only multi-class classification solvers are available. " \
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:76]:
```suggestion
            raise ValueError("You should use only classification solvers in LIBLINEAR. " \
```

File: libmultilabel/linear/linear.py [hunk 2]
```diff
@@ -66,23 +68,29 @@ def prepare_options(x: sparse.csr_matrix, options: str) -> 'tuple[sparse.csr_mat
...
+                "Please see LIBLINEAR README (https://github.com/cjlin1/liblinear) for details.")
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:77]: this line isn't needed

File: libmultilabel/linear/linear.py [hunk 3]
```diff
@@ -475,6 +483,22 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
     return {'weights': np.asmatrix(weights), '-B': bias, 'threshold': 0}
 
 
+def train_binary_and_multiclass(y: sparse.csr_matrix, x: sparse.csr_matrix, options: str):
+    """Trains a linear model for binary and multiclass data. See user guide for more details.
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:487]:
```suggestion
    """Trains a linear model for binary and multi-class data.
```

File: libmultilabel/linear/linear.py [hunk 4]
```diff
@@ -485,20 +509,28 @@ def predict_values(model, x: sparse.csr_matrix) -> np.ndarray:
...
-            np.zeros((x.shape[0], num_feature - x.shape[1])),
-            bias_col,
-        ], 'csr')
+    if isinstance(model, liblinear_model):
+        pred_labels, _, _ = predict([], x, model)
+        num_class = len(model.get_labels())
+        preds = np.zeros((x.shape[0], num_class))
+        ind = np.array(pred_labels, dtype='int')[..., np.newaxis]
+        np.put_along_axis(preds, ind, 1, axis=1)
+        return preds
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:518]: should return values. Should call predict_values of liblinear

File: libmultilabel/linear/metrics.py [hunk 5]
```diff
@@ -49,17 +49,24 @@ def compute(self) -> float:
...
             raise ValueError('unsupported average')
         self.average = average
+        self.multiclass = multiclass
         self.tp = self.fp = self.fn = 0
 
     def update(self, preds: np.ndarray, target: np.ndarray) -> None:
         assert preds.shape == target.shape  # (batch_size, num_classes)
-        preds = preds > self.metric_threshold
+        if self.multiclass:
+            # https://github.com/cjlin1/liblinear/blob/master/linear.cpp#L3333
```
Comment (cjlin1) [inline libmultilabel/linear/metrics.py:64]: I don't think this comment is needed

File: libmultilabel/linear/linear.py [hunk 6]
```diff
@@ -66,23 +58,28 @@ def prepare_options(x: sparse.csr_matrix, options: str) -> 'tuple[sparse.csr_mat
     if any(o in options for o in ['-R', '-C', '-v']):
         raise ValueError('-R, -C and -v are not supported')
 
+    options_split = options.split()
+    if '-s' in options_split:
+        i = options_split.index('-s')
+        solver_type = int(options_split[i+1])
+        if solver_type >= 11:
```
Comment (Sinacam) [inline libmultilabel/linear/linear.py:65]: Should it be `solver_type > 7`?

File: libmultilabel/linear/linear.py [hunk 7]
```diff
@@ -66,23 +58,28 @@ def prepare_options(x: sparse.csr_matrix, options: str) -> 'tuple[sparse.csr_mat
...
+            raise ValueError("You should use only classification solvers in LIBLINEAR.")
```
Comment (Sinacam) [inline libmultilabel/linear/linear.py:66]: Perhaps reword as "Invalid solver type, only classification solvers are allowed"?

File: libmultilabel/linear/linear.py [hunk 8]
```diff
@@ -138,7 +135,7 @@ def thresholding_one_label(y: np.ndarray,
 
     nr_fold = 3
 
-    l = y.shape[0]
+    l = y.shape[0]  # number of instances
```
Comment (Sinacam) [inline libmultilabel/linear/linear.py:138]: You should rename variable if you find the need to add such a comment.
That said, `l` has been the symbol for number of instances in our papers for a while now, which is why it's used without explanation.

File: libmultilabel/linear/linear.py [hunk 9]
```diff
@@ -250,35 +246,44 @@ def scutfbr(y: np.ndarray,
...
 
     Returns:
-        np.matrix: the weights.
+        dict: A dict of model weights and labels.
     """
     with silent_stderr():
         model = train(y, x, options)
 
-    w = np.ctypeslib.as_array(model.w, (x.shape[1], 1))
+    labels = set(y)
```
Comment (Sinacam) [inline libmultilabel/linear/linear.py:269]: Use `model.get_labels()` instead?

File: libmultilabel/linear/linear.py [hunk 10]
```diff
@@ -469,12 +475,38 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
...
+        options (str): The option string passed to liblinear.
+
+    Returns:
+        A model which can be used in predict_values.
+    """
+    x, options, bias = prepare_options(x, options)
+    y = np.squeeze(y.nonzero()[1])
+
+    model = do_train(y, x, options)
+    labels = model["labels"]
```
Comment (Sinacam) [inline libmultilabel/linear/linear.py:499]: Inconsistent double quotes.

File: libmultilabel/linear/metrics.py [hunk 11]
```diff
@@ -49,17 +49,23 @@ def compute(self) -> float:
...
             raise ValueError('unsupported average')
         self.average = average
+        self.multiclass = multiclass
         self.tp = self.fp = self.fn = 0
 
     def update(self, preds: np.ndarray, target: np.ndarray) -> None:
         assert preds.shape == target.shape  # (batch_size, num_classes)
-        preds = preds > self.metric_threshold
+        if self.multiclass:
+            max_idx = np.argmax(preds, axis=1).reshape(preds.shape[0], 1)
```
Comment (Sinacam) [inline libmultilabel/linear/metrics.py:64]: Perhaps `reshape(-1, 1)`?

File: libmultilabel/linear/linear.py (hunk 6 above)
Comment (Eleven1Liu) [reply libmultilabel/linear/linear.py:65]: Yes, this can be 7 not 11..

**Comment by Eleven1Liu** [comment]
after PR #231

File: libmultilabel/linear/linear.py [hunk 12]
```diff
@@ -35,16 +36,7 @@ def train_1vsrest(y: sparse.csr_matrix, x: sparse.csr_matrix, options: str):
...
-        w = np.ctypeslib.as_array(modeli.w, (num_feature,))
-        # Liblinear flips +1/-1 labels so +1 is always the first label,
-        # but not if all labels are -1.
-        # For our usage, we need +1 to always be the first label,
-        # so the check is necessary.
-        if modeli.get_labels()[0] == -1:
-            weights[:, i] = -w
-        else:
-            weights[:, i] = w
+        weights[:, i] = do_train(2*yi - 1, x, options).reshape(-1)
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:39]: why do we need reshape here?

File: libmultilabel/linear/linear.py (hunk 7 above)
Comment (cjlin1) [reply libmultilabel/linear/linear.py:66]: Let's use 
Invalid LIBLINEAR solver type. Only classification solvers are allowed

File: libmultilabel/linear/linear.py [hunk 13]
```diff
@@ -475,6 +472,43 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
...
+        y (sparse.csr_matrix): A 0/1 matrix with dimensions number of instances * number of classes.
+        x (sparse.csr_matrix): A matrix with dimensions number of instances * number of features.
+        options (str): The option string passed to liblinear.
+
+    Returns:
+        A model which can be used in predict_values.
+    """
+    x, options, bias = prepare_options(x, options)
+    num_labels = y.shape[1]
+    y = np.squeeze(y.nonzero()[1])
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:488]: I think Li-Chung mentioned we should check if each instance has exactly one label? If not, the resulting y may not be correct (longer or shorter than the # of data)

File: libmultilabel/linear/linear.py [hunk 14]
```diff
@@ -475,6 +472,43 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
...
+
+    with silent_stderr():
+        model = train(y, x, options)
+
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:492]: add a comment to explain train_label:
Number of labels appeared in training set; length may be smaller than num_labels

File: libmultilabel/linear/linear.py [hunk 15]
```diff
@@ -475,6 +472,43 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
...
+    train_labels = model.get_labels()
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:493]: this and next lines can be combined
train_labels = np.array(...
"ind" not a good name. Let's just use train_labels subsequently

File: libmultilabel/linear/linear.py [hunk 16]
```diff
@@ -475,6 +472,43 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
...
+    ind = np.array(train_labels, dtype='int')
+    weights = np.zeros((x.shape[1], num_labels))
+    if num_labels == 2 and '-s 4' not in options:
+        # For binary classification, liblinear returns weights
+        # with shape (number of instances * 1) except '-s 4'.
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:498]: number of features * 1

File: libmultilabel/linear/linear.py [hunk 17]
```diff
@@ -475,6 +472,43 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
...
+        w = np.ctypeslib.as_array(model.w, (x.shape[1], 1))
+        weights[:, ind[0]] = w[:, 0]
+        weights[:, ind[1]] = -w[:, 0]
+    else:
+        # Map label to the original index
+        w = np.ctypeslib.as_array(model.w, (x.shape[1], num_labels))
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:504]: I think a bug here.. Shouldn't use num_labels. Should be len(train_labels).

File: libmultilabel/linear/linear.py [hunk 18]
```diff
@@ -475,6 +472,43 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
...
+        weights[:, ind] = w
+
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:506]: Add a comment:
For labels not appeared in training, assign thresholds to -inf so they won't be predicted.

File: libmultilabel/linear/linear.py [hunk 19]
```diff
@@ -475,6 +472,45 @@ def train_cost_sensitive_micro(y: sparse.csr_matrix, x: sparse.csr_matrix, optio
...
+    """
+    x, options, bias = prepare_options(x, options)
+    num_instances, num_labels = y.shape
+    y = np.squeeze(y.nonzero()[1])
+    assert len(y) == num_instances, 'Invalid dataset. Only multi-class dataset is allowed.'
+
+    with silent_stderr():
+        model = train(y, x, options)
+
+    # Number of labels appeared in training set; length may be smaller than num_labels
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:494]:
```suggestion
    # Labels appeared in training set; length may be smaller than num_labels
```

**Review by cjlin1 — APPROVED** [review]

**Pull request merged**